### ISSUE TITLE
Browser/HTML  specify figure size rather than figsize * dpi

### ISSUE THREAD

**Issue by dansteingart** [mpld3/mpld3#65]
Really enjoying the package, many thanks for sharing. 

This is a suggested improvement, but it's not quite pythonic.  But then again neither is d3js :)

It's great that the figure's set size is reflected on screen, but I think it would be "more responsive" to have an option to set the canvas size to the bounding div, and then letting that div size be determine by the browser width/whatever else the developer wants.

Is it worth me mucking about in _js.py and _object.py to get this going?

**Comment by eadygol** [comment]
Hi, thanks for the great project. Has there been any progress on this issue? Has the plugin / feature been developed? Specifically, I am referring to "option to set the canvas size to the bounding div." Thanks.

**Comment by carlinmack** [comment]
I think this is a worthy feature, but for reference after plotting you can use d3 to make the figure autoscale (here with the window height)

```js
fetch("d3/" + figure + ".json")
    .then((response) => response.json())
    .then((data) => mpld3.draw_figure(figure, data))
    .then(() => {
        d3.select("#" + figure)
            .selectAll(".mpld3-figure")
            .attr("width", "100%")
            .attr("height", "70vh")
            .attr("viewBox", "0 0 1200 800")
            .attr("preserveAspectRatio", "xMinYMin meet");
    });

```

**Comment by perrette** [comment]
Here I'll second @hadim and @carlinmack to use `viewBox`. It can be very simple actually. For instance, if you currently have  `<svg width="1200" height="800">`, try instead `<svg viewBox="0 0 1200 800">. That's it. The `viewBox` defines the relative size of the svg, and the `width` and `height` parameters are not needed anymore. Well, just like a `div`, they can be set to say, `width=100%`, or be set in css directly (the css overwrites these values anyway).

I used that successfully with a recent chrome version. Cross-browser dependency should be investigated for the details, but I think the bottom line is clear. 

PS: there is a background article at https://css-tricks.com/scale-svg but I find it confusing when the method I outlined above just works (at least in modern browsers) and is simple conceptually: `viewBox = "0 0 ${width} ${height}"` for the inner code, `width=...` and `height=...` for the re-scaling, if necessary.

**Comment by perrette** [comment]
Yeah well, that's exactly what @carlinmack suggested. I was confused by "use d3 to make the figure autoscale", since it is pure HTML / CSS, but ok, manipulating the DOM is what d3 does. The "preserveAspectRatio" as indicated by @carlinmack is the default value I believe (at least that's what is explained in the css-tricks.com link). Why would you indicate `height="70vh"` if the aspect ratio is preserved anyway ? Is that some kind of upper bound for very narrow and tall figures ?

**Comment by carlinmack** [comment]
correct, I didn't want portrait figures being higher than the viewport :)

**Comment by perrette** [comment]
Except that it makes the figure bigger than necessary, not appropriate when a tight fit with surrounding content is needed. Using `max-height` instead works for me.

Here is how I end up loading mpld3 figures in the browser:

```javascript
      let figureID = "id_12345" ; 
      let dataURL = figureID + ".json" ; 

      d3.json(dataURL).then( function(data) {
        mpld3.draw_figure(figureID, data);
        return data;
      }).then( function(data) {
        d3.select("#" + figureID)
          .selectAll(".mpld3-figure")
          .attr("width", null)   // remove "width" and "height" attributes as set by mpld3
          .attr("height", null)  
          .attr("viewBox", `0 0 ${data.width} ${data.height}`)
      });
```

**Comment by cliffckerr** [comment]
Thanks for the discussion @carlinmack @perrette ! Coincidentally we just encountered this too, @dkong-idm's solution was:
```
	const resizeChart = (viewBoxSettings) => {
		let svg = document.getElementsByClassName("mpld3-figure");

		if (svg && svg.length > 0) {
			svg[0].setAttribute("viewBox",viewBoxSettings );
			svg[0].setAttribute("width", "90vw");
			svg[0].setAttribute("height", "auto");
		}
	}
```
@vladh , would it be a lot of work to add this auto-scaling functionality to `mpld3`, either as an optional arg to `draw_figure` or else as a standalone function?

**Comment by vladh** [comment]
@cliffckerr I can definitely add this to `draw_figure`, perhaps as an argument that takes an `Object` and sets all the attributes specified in that object on the `svg`, so you could pass in something like `{viewBox: ..., width: '90vw', height: 'auto'}`.

My question is if you think this should be added anywhere else, to support e.g. `fig_to_html()`.

**Comment by cbbcbail** [comment]
@vladh Yes, it would also be very useful in `fig_to_html()`.
